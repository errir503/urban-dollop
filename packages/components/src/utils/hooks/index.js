export { default as useControlledState } from './use-controlled-state';
export { default as useJumpStep } from './use-jump-step';
export { default as useUpdateEffect } from './use-update-effect';
export { useControlledValue } from './use-controlled-value';
export { useCx } from './use-cx';
export { useLatestRef } from './use-latest-ref';
