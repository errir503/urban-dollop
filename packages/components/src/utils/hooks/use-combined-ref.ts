/**
 * WordPress dependencies
 */
import { useRef, useEffect } from '@wordpress/element';
/**
 * External dependencies
 */
// eslint-disable-next-line no-restricted-imports
import type { MutableRefObject, RefCallback } from 'react';

type Ref< T > = MutableRefObject< T | null > | RefCallback< T | null >;

export function useCombinedRef< T extends HTMLElement >( ...refs: Ref< T >[] ) {
	const targetRef = useRef( null );

	useEffect( () => {
		refs.forEach( ( ref ) => {
			if ( ! ref ) return;

			if ( typeof ref === 'function' ) {
				ref( targetRef.current );
			} else {
				ref.current = targetRef.current;
			}
		} );
	}, [ refs ] );

	return targetRef;
}
