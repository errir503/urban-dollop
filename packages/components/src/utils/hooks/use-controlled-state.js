/**
 * WordPress dependencies
 */
import { useEffect, useState } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { isValueDefined, getDefinedValue } from '../values';

/**
 * @template T
 * @typedef Options
 * @property {T | undefined} initial  Initial value
 * @property {T | ""}        fallback Fallback value
 */

/** @type {Readonly<{ initial: undefined, fallback: '' }>} */
const defaultOptions = {
	initial: undefined,
	/**
	 * Defaults to empty string, as that is preferred for usage with
	 * <input />, <textarea />, and <select /> form elements.
	 */
	fallback: '',
};

/**
 * Custom hooks for "controlled" components to track and consolidate internal
 * state and incoming values. This is useful for components that render
 * `input`, `textarea`, or `select` HTML elements.
 *
 * https://reactjs.org/docs/forms.html#controlled-components
 *
 * At first, a component using useControlledState receives an initial prop
 * value, which is used as initial internal state.
 *
 * This internal state can be maintained and updated without
 * relying on new incoming prop values.
 *
 * Unlike the basic useState hook, useControlledState's state can
 * be updated if a new incoming prop value is changed.
 *
 * @template T
 *
 * @param {T | undefined} currentState             The current value.
 * @param {Options<T>}    [options=defaultOptions] Additional options for the hook.
 *
 * @return {[T | "", (nextState: T) => void]} The controlled value and the value setter.
 */
function useControlledState( currentState, options = defaultOptions ) {
	const { initial, fallback } = { ...defaultOptions, ...options };

	const [ internalState, setInternalState ] = useState( currentState );
	const hasCurrentState = isValueDefined( currentState );

	/*
	 * Resets internal state if value every changes from uncontrolled <-> controlled.
	 */
	useEffect( () => {
		if ( hasCurrentState && internalState ) {
			setInternalState( undefined );
		}
	}, [ hasCurrentState, internalState ] );

	const state = getDefinedValue(
		[ currentState, internalState, initial ],
		fallback
	);

	/* eslint-disable jsdoc/no-undefined-types */
	/** @type {(nextState: T) => void} */
	const setState = ( nextState ) => {
		if ( ! hasCurrentState ) {
			setInternalState( nextState );
		}
	};
	/* eslint-enable jsdoc/no-undefined-types */

	return [ state, setState ];
}

export default useControlledState;
