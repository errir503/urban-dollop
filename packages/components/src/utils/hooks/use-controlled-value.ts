/**
 * WordPress dependencies
 */
import { useState } from '@wordpress/element';

type Props< T > = {
	defaultValue?: T;
	value?: T;
	onChange?: ( value: T ) => void;
};

/**
 * Simplified and improved implementation of useControlledState.
 *
 * @param  props
 * @param  props.defaultValue
 * @param  props.value
 * @param  props.onChange
 * @return The controlled value and the value setter.
 */
export function useControlledValue< T >( {
	defaultValue,
	onChange,
	value: valueProp,
}: Props< T > ): [ T | undefined, ( value: T ) => void ] {
	const hasValue = typeof valueProp !== 'undefined';
	const initialValue = hasValue ? valueProp : defaultValue;
	const [ state, setState ] = useState( initialValue );
	const value = hasValue ? valueProp : state;

	let setValue: ( nextValue: T ) => void;
	if ( hasValue && typeof onChange === 'function' ) {
		setValue = onChange;
	} else if ( ! hasValue && typeof onChange === 'function' ) {
		setValue = ( nextValue ) => {
			onChange( nextValue );
			setState( nextValue );
		};
	} else {
		setValue = setState;
	}

	return [ value, setValue ];
}
