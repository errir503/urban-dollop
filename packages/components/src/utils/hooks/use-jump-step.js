/**
 * WordPress dependencies
 */
import { useEffect, useState } from '@wordpress/element';

/**
 * A custom hook that calculates a step value (used by elements like input
 * [type="number"]). This value can be modified based on whether the Shift
 * key is being held down.
 *
 * For example, a shiftStep of 10, and a step of 1...
 * Starting from 10, the next incremented value will be 11.
 *
 * Holding down shift...
 * Starting from 10, the next incremented value will be 20.
 *
 * @param {Object}  props                           Properties for the hook.
 * @param {boolean} [props.isShiftStepEnabled=true] Determines if jumping values with shift is enabled
 * @param {number}  [props.shiftStep=10]            Multiplier to jump by, when holding shift key.
 * @param {number}  [props.step=1]                  Multiplier to jump by, when not-holding shift key.
 *
 * @return {number} The jump step value.
 */
function useJumpStep( {
	isShiftStepEnabled = true,
	shiftStep = 10,
	step = 1,
} ) {
	const [ isShiftKey, setIsShiftKey ] = useState( false );

	useEffect( () => {
		/** @type {(event: KeyboardEvent)=>void} */
		const handleShiftKeyToggle = ( event ) => {
			setIsShiftKey( event.shiftKey );
		};

		window.addEventListener( 'keydown', handleShiftKeyToggle );
		window.addEventListener( 'keyup', handleShiftKeyToggle );

		return () => {
			window.removeEventListener( 'keydown', handleShiftKeyToggle );
			window.removeEventListener( 'keyup', handleShiftKeyToggle );
		};
	}, [] );

	const isEnabled = isShiftStepEnabled && isShiftKey;

	return isEnabled ? shiftStep * step : step;
}

export default useJumpStep;
