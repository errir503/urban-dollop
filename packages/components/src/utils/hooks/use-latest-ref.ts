/**
 * External dependencies
 */
import type { RefObject } from 'react';

/**
 * WordPress dependencies
 */
import { useRef } from '@wordpress/element';
import { useIsomorphicLayoutEffect } from '@wordpress/compose';

/**
 * Creates a reference for a prop. This is useful for preserving dependency
 * memoization for hooks like useCallback.
 *
 * @see https://codesandbox.io/s/uselatestref-mlj3i?file=/src/App.tsx
 *
 * @param  value The value to reference
 * @return The prop reference.
 */
export function useLatestRef< T >( value: T ): RefObject< T > {
	const ref = useRef( value );

	useIsomorphicLayoutEffect( () => {
		ref.current = value;
	} );

	return ref;
}
