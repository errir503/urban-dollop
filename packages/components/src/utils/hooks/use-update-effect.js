/**
 * WordPress dependencies
 */
import { useRef, useEffect } from '@wordpress/element';

/**
 * A `React.useEffect` that will not run on the first render.
 * Source:
 * https://github.com/reakit/reakit/blob/HEAD/packages/reakit-utils/src/useUpdateEffect.ts
 *
 * @param {import('react').EffectCallback} effect
 * @param {import('react').DependencyList} deps
 */
function useUpdateEffect( effect, deps ) {
	const mounted = useRef( false );

	useEffect( () => {
		if ( mounted.current ) {
			return effect();
		}
		mounted.current = true;
		return undefined;
	}, deps );
}

export default useUpdateEffect;
