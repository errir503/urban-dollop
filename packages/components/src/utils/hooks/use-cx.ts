/**
 * External dependencies
 */
// eslint-disable-next-line no-restricted-imports
import type { Context } from 'react';
import { CacheProvider, EmotionCache } from '@emotion/react';
import type { SerializedStyles } from '@emotion/serialize';
import { insertStyles } from '@emotion/utils';
// eslint-disable-next-line no-restricted-imports
import { cx as innerCx, ClassNamesArg } from '@emotion/css';

/**
 * WordPress dependencies
 */
import { useContext, useCallback } from '@wordpress/element';

// @ts-ignore Private property
const EmotionCacheContext: Context< EmotionCache > = CacheProvider._context;

const useEmotionCacheContext = () => useContext( EmotionCacheContext );

const isSerializedStyles = ( o: any ): o is SerializedStyles =>
	// eslint-disable-next-line eqeqeq
	o != null &&
	[ 'name', 'styles' ].every( ( p ) => typeof o[ p ] !== 'undefined' );

/**
 * Retrieve a `cx` function that knows how to handle `SerializedStyles`
 * returned by the `@emotion/react` `css` function in addition to what
 * `cx` normally knows how to handle. It also hooks into the Emotion
 * Cache, allowing `css` calls to work inside iframes.
 *
 * @example
 * import { css } from '@emotion/react';
 *
 * const styles = css`
 * 	color: red
 * `;
 *
 * function RedText( { className, ...props } ) {
 * 	const cx = useCx();
 *
 * 	const classes = cx(styles, className);
 *
 * 	return <span className={classes} {...props} />;
 * }
 */
export const useCx = () => {
	const cache = useEmotionCacheContext();

	const cx = useCallback(
		( ...classNames: ( ClassNamesArg | SerializedStyles )[] ) => {
			return innerCx(
				...classNames.map( ( arg ) => {
					if ( isSerializedStyles( arg ) ) {
						insertStyles( cache, arg, false );
						return `${ cache.key }-${ arg.name }`;
					}
					return arg;
				} )
			);
		},
		[ cache ]
	);

	return cx;
};
