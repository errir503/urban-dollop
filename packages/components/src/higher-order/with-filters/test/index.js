/**
 * External dependencies
 */
import { shallow } from 'enzyme';
import TestUtils from 'react-dom/test-utils';
import ReactDOM from 'react-dom';

/**
 * WordPress dependencies
 */
import { addFilter, removeAllFilters, removeFilter } from '@wordpress/hooks';
import { Component } from '@wordpress/element';

/**
 * Internal dependencies
 */
import withFilters from '..';

const assertExpectedHtml = ( wrapper, expectedHTML ) => {
	// eslint-disable-next-line react/no-find-dom-node
	const element = ReactDOM.findDOMNode( wrapper );
	expect( element.outerHTML ).toBe( expectedHTML );
};

// This is needed because TestUtils does not accept a stateless component.
// anything run through a HOC ends up as a stateless component.
const getTestComponent = ( WrappedComponent ) => {
	class TestComponent extends Component {
		render() {
			return <WrappedComponent { ...this.props } />;
		}
	}
	return <TestComponent />;
};

describe( 'withFilters', () => {
	let shallowWrapper, wrapper;

	const hookName = 'EnhancedComponent';
	const MyComponent = () => <div>My component</div>;

	afterEach( () => {
		if ( wrapper ) {
			ReactDOM.unmountComponentAtNode(
				// eslint-disable-next-line react/no-find-dom-node
				ReactDOM.findDOMNode( wrapper ).parentNode
			);
		}
		if ( shallowWrapper ) {
			shallowWrapper.unmount();
		}
		removeAllFilters( hookName );
	} );

	it( 'should display original component when no filters applied', () => {
		const EnhancedComponent = withFilters( hookName )( MyComponent );

		shallowWrapper = shallow( <EnhancedComponent /> );

		expect( shallowWrapper.html() ).toBe( '<div>My component</div>' );
	} );

	it( 'should display a component overridden by the filter', () => {
		const OverriddenComponent = () => <div>Overridden component</div>;
		addFilter(
			'EnhancedComponent',
			'test/enhanced-component-override',
			() => OverriddenComponent
		);
		const EnhancedComponent = withFilters( hookName )( MyComponent );

		shallowWrapper = shallow( <EnhancedComponent /> );

		expect( shallowWrapper.html() ).toBe(
			'<div>Overridden component</div>'
		);
	} );

	it( 'should display two components composed by the filter', () => {
		const ComposedComponent = () => <div>Composed component</div>;
		addFilter(
			hookName,
			'test/enhanced-component-compose',
			( FilteredComponent ) => () => (
				<div>
					<FilteredComponent />
					<ComposedComponent />
				</div>
			)
		);
		const EnhancedComponent = withFilters( hookName )( MyComponent );

		shallowWrapper = shallow( <EnhancedComponent /> );

		expect( shallowWrapper.html() ).toBe(
			'<div><div>My component</div><div>Composed component</div></div>'
		);
	} );

	it( 'should not re-render component when new filter added before component was mounted', () => {
		const spy = jest.fn();
		const SpiedComponent = () => {
			spy();
			return <div>Spied component</div>;
		};
		addFilter(
			hookName,
			'test/enhanced-component-spy-1',
			( FilteredComponent ) => () => (
				<blockquote>
					<FilteredComponent />
				</blockquote>
			)
		);
		const EnhancedComponent = withFilters( hookName )( SpiedComponent );

		wrapper = TestUtils.renderIntoDocument(
			getTestComponent( EnhancedComponent )
		);

		jest.runAllTimers();

		expect( spy ).toHaveBeenCalledTimes( 1 );
		assertExpectedHtml(
			wrapper,
			'<blockquote><div>Spied component</div></blockquote>'
		);
	} );

	it( 'should re-render component once when new filter added after component was mounted', () => {
		const spy = jest.fn();
		const SpiedComponent = () => {
			spy();
			return <div>Spied component</div>;
		};
		const EnhancedComponent = withFilters( hookName )( SpiedComponent );

		wrapper = TestUtils.renderIntoDocument(
			getTestComponent( EnhancedComponent )
		);

		spy.mockClear();
		addFilter(
			hookName,
			'test/enhanced-component-spy-1',
			( FilteredComponent ) => () => (
				<blockquote>
					<FilteredComponent />
				</blockquote>
			)
		);
		jest.runAllTimers();

		expect( spy ).toHaveBeenCalledTimes( 1 );
		assertExpectedHtml(
			wrapper,
			'<blockquote><div>Spied component</div></blockquote>'
		);
	} );

	it( 'should re-render component once when two filters added in the same animation frame', () => {
		const spy = jest.fn();
		const SpiedComponent = () => {
			spy();
			return <div>Spied component</div>;
		};
		const EnhancedComponent = withFilters( hookName )( SpiedComponent );
		wrapper = TestUtils.renderIntoDocument(
			getTestComponent( EnhancedComponent )
		);

		spy.mockClear();

		addFilter(
			hookName,
			'test/enhanced-component-spy-1',
			( FilteredComponent ) => () => (
				<blockquote>
					<FilteredComponent />
				</blockquote>
			)
		);
		addFilter(
			hookName,
			'test/enhanced-component-spy-2',
			( FilteredComponent ) => () => (
				<section>
					<FilteredComponent />
				</section>
			)
		);
		jest.runAllTimers();

		expect( spy ).toHaveBeenCalledTimes( 1 );
		assertExpectedHtml(
			wrapper,
			'<section><blockquote><div>Spied component</div></blockquote></section>'
		);
	} );

	it( 'should re-render component twice when new filter added and removed in two different animation frames', () => {
		const spy = jest.fn();
		const SpiedComponent = () => {
			spy();
			return <div>Spied component</div>;
		};
		const EnhancedComponent = withFilters( hookName )( SpiedComponent );
		wrapper = TestUtils.renderIntoDocument(
			getTestComponent( EnhancedComponent )
		);

		spy.mockClear();
		addFilter(
			hookName,
			'test/enhanced-component-spy',
			( FilteredComponent ) => () => (
				<div>
					<FilteredComponent />
				</div>
			)
		);
		jest.runAllTimers();

		removeFilter( hookName, 'test/enhanced-component-spy' );
		jest.runAllTimers();

		expect( spy ).toHaveBeenCalledTimes( 2 );
		assertExpectedHtml( wrapper, '<div>Spied component</div>' );
	} );

	it( 'should re-render both components once each when one filter added', () => {
		const spy = jest.fn();
		const SpiedComponent = () => {
			spy();
			return <div>Spied component</div>;
		};
		const EnhancedComponent = withFilters( hookName )( SpiedComponent );
		const CombinedComponents = () => (
			<section>
				<EnhancedComponent />
				<EnhancedComponent />
			</section>
		);
		wrapper = TestUtils.renderIntoDocument(
			getTestComponent( CombinedComponents )
		);

		spy.mockClear();
		addFilter(
			hookName,
			'test/enhanced-component-spy-1',
			( FilteredComponent ) => () => (
				<blockquote>
					<FilteredComponent />
				</blockquote>
			)
		);
		jest.runAllTimers();

		expect( spy ).toHaveBeenCalledTimes( 2 );
		assertExpectedHtml(
			wrapper,
			'<section><blockquote><div>Spied component</div></blockquote><blockquote><div>Spied component</div></blockquote></section>'
		);
	} );
} );
