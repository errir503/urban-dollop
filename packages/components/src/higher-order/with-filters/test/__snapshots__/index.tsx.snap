// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withFilters should display a component overridden by the filter 1`] = `
<div>
  <div>
    Overridden component
  </div>
</div>
`;

exports[`withFilters should display original component when no filters applied 1`] = `
<div>
  <div>
    My component
  </div>
</div>
`;

exports[`withFilters should display two components composed by the filter 1`] = `
<div>
  <div>
    <div>
      My component
    </div>
    <div>
      Composed component
    </div>
  </div>
</div>
`;

exports[`withFilters should not re-render component when new filter added before component was mounted 1`] = `
<div>
  <blockquote>
    <div>
      Spied component
    </div>
  </blockquote>
</div>
`;

exports[`withFilters should re-render both components once each when one filter added 1`] = `
<div>
  <section>
    <blockquote>
      <div>
        Spied component
      </div>
    </blockquote>
    <blockquote>
      <div>
        Spied component
      </div>
    </blockquote>
  </section>
</div>
`;

exports[`withFilters should re-render component once when new filter added after component was mounted 1`] = `
<div>
  <blockquote>
    <div>
      Spied component
    </div>
  </blockquote>
</div>
`;

exports[`withFilters should re-render component once when two filters added in the same animation frame 1`] = `
<div>
  <section>
    <blockquote>
      <div>
        Spied component
      </div>
    </blockquote>
  </section>
</div>
`;

exports[`withFilters should re-render component twice when new filter added and removed in two different animation frames 1`] = `
<div>
  <div>
    Spied component
  </div>
</div>
`;
